state-transition/core/state_processor_withdrawals.go

```diff
@@ -136,3 +155,186 @@ func (sp *StateProcessor) processWithdrawals(
+
+	hasExcess := balance > minActivationBalance+pendingBalanceToWithdraw
+
+	if validator.HasCompoundingWithdrawalCredential() && hasSufficient && hasExcess {
+		toWithdraw := min(balance-minActivationBalance-pendingBalanceToWithdraw, req.Amount)
+		// As long as `processPartialWithdrawal` is called after `processSlots`, this will always return the correct slot.
+		currentSlot, getErr := st.GetSlot()
+		if getErr != nil {
+			return getErr
+		}
+		nextEpoch := sp.cs.SlotToEpoch(currentSlot) + 1
+		// Note that we do not need to set the ExitEpoch anywhere here as Partial withdrawals do not exit the validator
+		// as we enforce that the `toWithdraw` amount is always above the `minActivationBalance`.
+		// For example, if a validator's balance is already at `minActivationBalance` (250K BERA on mainnet)
+		// and they request a partial withdrawal, the `toWithdraw` amount will always be zero.
+		withdrawableEpoch := math.Epoch(uint64(nextEpoch) + sp.cs.MinValidatorWithdrawabilityDelay())
+		ppWithdrawal := &ctypes.PendingPartialWithdrawal{
+			ValidatorIndex:    index,
+			Amount:            toWithdraw,
+			WithdrawableEpoch: withdrawableEpoch,
+		}
+		pendingWithdrawals = append(pendingWithdrawals, ppWithdrawal)
+		return st.SetPendingPartialWithdrawals(pendingWithdrawals)
+	}
+	return nil
+}
+
+// validateWithdrawal checks that the validator exists and that the withdrawal credentials match.
+func validateWithdrawal(st *state.StateDB, withdrawalRequest *ctypes.WithdrawalRequest) (math.ValidatorIndex, *ctypes.Validator, error) {
+	// Verify pubkey exists
+	index, err := st.ValidatorIndexByPubkey(withdrawalRequest.ValidatorPubKey)
+	if err != nil {
+		return 0, nil, err
+	}
+
+	validator, err := st.ValidatorByIndex(index)
+	if err != nil {
+		return 0, nil, err
+	}
+	if validator == nil {
+		// This should never occur as we expect ErrNotFound if the validator is not found.
+		return 0, nil, errors.New("validateWithdrawal: validator does not exist")
+	}
+
+	// Verify withdrawal credentials
+	if !validator.HasExecutionWithdrawalCredential() {
+		return 0, nil, errors.New("validator does not have execution withdrawal credentials")
+	}
+	correctCred, err := validator.WithdrawalCredentials.ToExecutionAddress()
+	if err != nil {
+		return 0, nil, err
+	}
+	if !withdrawalRequest.SourceAddress.Equals(correctCred) {
+		return 0, nil, errors.New("source address does not match execution withdrawal credential")
+	}
+	return index, validator, nil
+}
+
+// verifyWithdrawalConditions checks additional conditions like active status, exit not initiated, and minimal activation period.
+func verifyWithdrawalConditions(chainSpec ChainSpec, st *state.StateDB, validator *ctypes.Validator) error {
+	slot, err := st.GetSlot()
+	if err != nil {
+		return err
+	}
+	currentEpoch := chainSpec.SlotToEpoch(slot)
+	// Verify the validator is active
+	if !validator.IsActive(currentEpoch) {
```

Comment:
not for myself: this covers the case where a validator eviction happens while a withdrawal request is being submitted. It does not cover the case where the withdrawal request is already accepted and the validator is evicted before the request is processed.